packages/builtin/string.pony

```diff
@@ -93,56 +93,39 @@ class val String is (Seq[U8] & Comparable[String box] & Stringable)
     """
     Create a UTF-8 string from a single UTF-32 code point.
     """
-    if value < 0x80 then
-      _size = 1
-      _alloc = _size + 1
-      _ptr = Pointer[U8]._alloc(_alloc)
-      _set(0, value.u8())
-    elseif value < 0x800 then
-      _size = 2
-      _alloc = _size + 1
-      _ptr = Pointer[U8]._alloc(_alloc)
-      _set(0, ((value >> 6) or 0xC0).u8())
-      _set(1, ((value and 0x3F) or 0x80).u8())
-    elseif value < 0xD800 then
-      _size = 3
-      _alloc = _size + 1
-      _ptr = Pointer[U8]._alloc(_alloc)
-      _set(0, ((value >> 12) or 0xE0).u8())
-      _set(1, (((value >> 6) and 0x3F) or 0x80).u8())
-      _set(2, ((value and 0x3F) or 0x80).u8())
-    elseif value < 0xE000 then
-      // UTF-16 surrogate pairs are not allowed.
-      _size = 3
-      _alloc = _size + 1
-      _ptr = Pointer[U8]._alloc(_alloc)
-      _set(0, 0xEF)
-      _set(1, 0xBF)
-      _set(2, 0xBD)
-      _size = _size + 3
-    elseif value < 0x10000 then
-      _size = 3
-      _alloc = _size + 1
-      _ptr = Pointer[U8]._alloc(_alloc)
-      _set(0, ((value >> 12) or 0xE0).u8())
-      _set(1, (((value >> 6) and 0x3F) or 0x80).u8())
-      _set(2, ((value and 0x3F) or 0x80).u8())
-    elseif value < 0x110000 then
-      _size = 4
-      _alloc = _size + 1
-      _ptr = Pointer[U8]._alloc(_alloc)
-      _set(0, ((value >> 18) or 0xF0).u8())
-      _set(1, (((value >> 12) and 0x3F) or 0x80).u8())
-      _set(2, (((value >> 6) and 0x3F) or 0x80).u8())
-      _set(3, ((value and 0x3F) or 0x80).u8())
-    else
-      // Code points beyond 0x10FFFF are not allowed.
-      _size = 3
-      _alloc = _size + 1
-      _ptr = Pointer[U8]._alloc(_alloc)
-      _set(0, 0xEF)
-      _set(1, 0xBF)
-      _set(2, 0xBD)
+    let encoded = _UTF32Encoder.encode(value)
+    _size = encoded._1
+    _alloc = _size + 1
+    _ptr = Pointer[U8]._alloc(_alloc)
+    _set(0, encoded._2)
+    if encoded._1 > 1 then
+      _set(1, encoded._3)
+      if encoded._1 > 2 then
+        _set(2, encoded._4)
+        if encoded._1 > 3 then
+          _set(3, encoded._5)
+        end
+      end
+    end
+    _set(_size, 0)
+
+  fun ref push_utf32(value: U32) =>
+    """
+    Push a UTF-32 code point.
+    """
+    let encoded = _UTF32Encoder.encode(value)
+    let i = _size
+    _size = _size + encoded._1
+    reserve(_size)
+    _set(i, encoded._2)
+    if encoded._1 > 1 then
+      _set(i + 1, encoded._3)
+      if encoded._1 > 2 then
+        _set(i + 2, encoded._4)
+        if encoded._1 > 3 then
+          _set(i + 3, encoded._5)
+        end
+      end
     end
```

Comment:
I don't know how `from_utf32` is used so I tried to make as little change as possible, especially regarding the memory footprint. `String.push_utf32(0x16DD)` will have an additional memory reallocation compared to `String. from_utf32(0x16DD)`. If you think that it is negligible, the code could be way more simpler indeed, without necessarily breaking the API by removing the constructor `from_utf32`.